File: typescript/preliminar/src/app/common/errors/either.ts
export type Either<L, R> = Left<L, R> | Right<L, R>;

export class Left<L, R> {
  private _value: L;

  constructor(value: L) {
    this._value = value;
  }

  isLeft(): this is Left<L, R> {
    return true;
  }

  isRight(): this is Right<L, R> {
    return false;
  }

  get value() {
    return this._value;
  }
}

export class Right<L, R> {
  private _value: R;

  constructor(value: R) {
    this._value = value;
  }

  isLeft(): this is Left<L, R> {
    return false;
  }

  isRight(): this is Right<L, R> {
    return true;
  }

  get value() {
    return this._value;
  }
}

export const left = <L, R>(l: L): Either<L, R> => {
  return new Left(l);
};

export const right = <L, R>(r: R): Either<L, R> => {
  return new Right(r);
};

File: python/preliminar/src/app/common/errors/either.py
from typing import Generic, TypeVar, Union

L = TypeVar('L')
R = TypeVar('R')

class Either(Generic[L, R]):
    def __init__(self, value: Union[L, R]):
        self._value = value

    def is_left(self) -> bool:
        return isinstance(self, Left)

    def is_right(self) -> bool:
        return isinstance(self, Right)

    @property
    def value(self) -> Union[L, R]:
        return self._value

class Left(Either):
    def __init__(self, value: L):
        super().__init__(value)

class Right(Either):
    def __init__(self, value: R):
        super().__init__(value)
